tla-bmcmt/src/main/scala/at/forsyte/apalache/tla/bmcmt/rules/FunExceptRuleWithArrays.scala

```diff
@@ -0,0 +1,73 @@
+package at.forsyte.apalache.tla.bmcmt.rules
+
+import at.forsyte.apalache.tla.bmcmt._
+import at.forsyte.apalache.tla.lir.TypedPredefs.{BuilderExAsTyped, tlaExToBuilderExAsTyped}
+import at.forsyte.apalache.tla.lir.convenience._
+import at.forsyte.apalache.tla.lir.{BoolT1, FunT1, TlaEx, TupT1}
+import at.forsyte.apalache.tla.lir.UntypedPredefs._
+
+/**
+ * Rewriting EXCEPT for functions, tuples, and records.
+ *
+ * @author
+ *   Rodrigo Otoni
+ */
+class FunExceptRuleWithArrays(rewriter: SymbStateRewriter) extends FunExceptRule(rewriter) {
+
+  // TODO: override rewriteRec and rewriteTuple later
+
+  override def rewriteFun(state: SymbState, funCell: ArenaCell, funT: FunT1, indexCell: ArenaCell,
+      valueCell: ArenaCell): SymbState = {
+
+    var nextState = state.updateArena(_.appendCell(funCell.cellType))
+    val resultFunCell = nextState.arena.topCell
+
+    // Propagate the function's domain
+    val domainCell = nextState.arena.getDom(funCell)
+    nextState = nextState.updateArena(_.setDom(resultFunCell, domainCell))
+
+    // Make pair <arg,res> to propagate metadata
+    def mkPair(indexCell: ArenaCell, resCell: ArenaCell): TlaEx = {
+      tla.tuple(indexCell.toNameEx, resCell.toNameEx).typed(TupT1(funT.arg, funT.res))
+    }
+
+    nextState = rewriter.rewriteUntilDone(nextState.setRex(mkPair(indexCell, valueCell)))
+    val newPairCell = nextState.asCell
+
+    // Declare the updated set of pairs <arg,res>
+    val relation = nextState.arena.getCdm(funCell)
+    val relationCells = nextState.arena.getHas(relation)
+    nextState = nextState.updateArena(_.appendCellNoSmt(relation.cellType))
+    val resultRelation = nextState.arena.topCell
+
+    def eachRelationPair(pair: ArenaCell) = {
+      val tupT = TupT1(funT.arg, funT.res)
+      val pairIndex = nextState.arena.getHas(pair).head
+      val ite = tla
+        .ite(tla.eql(pairIndex.toNameEx as tupT, indexCell.toNameEx as funT.arg) as BoolT1(),
+            newPairCell.toNameEx as tupT, pair.toNameEx as tupT)
+        .typed(Map("p" -> tupT, "i" -> funT.arg, "b" -> BoolT1()), "p")
```

Comment:
```suggestion
        as BoolT1()
```

Reply:
The type here is not bool but tup, so this doesn't work.